interpreter/src/main/scala/pl/touk/nussknacker/engine/compile/nodecompilation/BaseNodeCompiler.scala

```diff
@@ -0,0 +1,325 @@
+        .valueOr(err => ExpressionCompilation(fieldName, None, Invalid(err)))
+
+      val nodeCompilation: NodeCompilationResult[expression.Expression] = NodeCompilationResult(
+        expressionTypingInfo = expressionCompilation.expressionTypingInfo,
+        parameters = None,
+        validationContext =
+          outputVar.map(ctx.withVariable(_, expressionCompilation.typingResult)).getOrElse(Valid(ctx)),
+        compiledObject = expressionCompilation.validated,
+        expressionType = Some(expressionCompilation.typingResult)
+      )
+
+      (expressionCompilation, nodeCompilation)
+    }
+
+  }
+
+  private object RecordValidator {
+
+    case class IndexedField(key: String, value: TypedExpression, index: Int)
+
+    def validateRecord(
+        fields: List[IndexedField]
+    )(implicit nodeId: NodeId): ValidatedNel[ProcessCompilationError, Unit] = {
+      fields.map { field =>
+        (
+          ValidationAdapter.validateMaybeVariable(Some(field.value), recordValueFieldName(field.index)),
+          validateUniqueRecordKey(fields.map(_.key), field.key, recordKeyFieldName(field.index))
+        ).mapN { (_, _) => () }
+      }.combineAll
+    }
+
+    private def validateUniqueRecordKey(allValues: Seq[String], requiredToBeUniqueValue: String, fieldName: String)(
+        implicit nodeId: NodeId
+    ) = {
+      if (allValues.count(_ == requiredToBeUniqueValue) <= 1) {
+        valid(())
+      } else
+        {
+          Invalid(
+            CustomParameterValidationError(
+              "The key of a record has to be unique",
+              "Record key not unique",
+              fieldName,
+              nodeId.id
+            )
+          )
+        }.toValidatedNel
+    }
+
+  }
+
+  private object ValidationAdapter {
+
+    def validateMaybeBoolean(
+        expression: Option[TypedExpression],
+        fieldName: String
+    )(
+        implicit nodeId: NodeId
+    ): Validated[NonEmptyList[PartSubGraphCompilationError], Unit] = {
+      validateOrValid(Typed[Boolean], NotNullParameterValidator, expression, fieldName)
+    }
+
+    def validateMaybeVariable(
+        expression: Option[TypedExpression],
+        fieldName: String
+    )(
+        implicit nodeId: NodeId
+    ): Validated[NonEmptyList[PartSubGraphCompilationError], Unit] = {
+      validateOrValid(Unknown, MandatoryParameterValidator, expression, fieldName)
+    }
+
+    private def validateOrValid(
+        expectedType: TypingResult,
+        validator: ParameterValidator,
+        expression: Option[TypedExpression],
+        fieldName: String
+    )(implicit nodeId: NodeId): Validated[NonEmptyList[PartSubGraphCompilationError], Unit] = {
+      expression
+        .map { expr =>
+          Validations
+            // The parameter definition name does not mean anything here
+            .validate(Parameter("stub", expectedType, List(validator)), (TypedParameter(fieldName, expr), ()))
```

Comment:
Why not to just extract another method in `Validations` instead of creating fake parameter?

Reply:
done